File: 백준/Silver/2108. 통계학/통계학.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder print = new StringBuilder();
    static int N, max, min, input, sum;
    static int[] numbers;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());
        max = -4000;
        min = 4000;
        sum = 0;
        numbers = new int[N];
        for (int i = 0; i < N; i++) {
            input = Integer.parseInt(br.readLine());
            sum += input;
            if (input > max) max = input;
            if (input < min) min = input;
            numbers[i] = input;
        }
        print.append(Math.round(sum / (double) N)).append('\n');
        Arrays.sort(numbers);
        print.append(numbers[N / 2]).append('\n');
        print.append(mode()).append('\n');
        print.append(max - min);
        System.out.println(print);
    }
    static int mode() {
        int k = max - min + 1;
        int[] array = new int[k];
        if (numbers.length == 1) return numbers[0];
        for (int i:numbers) {
            array[i - min]++;
        }
        int min1 = 4001, min2 = 4001, frequency = 1;
        for (int i = 0; i < k; i++) {
            if (array[i] > frequency) {
                frequency = array[i];
                min1 = i + min;
                min2 = 4001;
            } else if (array[i] == frequency) {
                if (i + min < min1) {
                    min2 = min1;
                    min1 = i + min;
                }
                else if (i + min < min2) min2 = i + min;
            }
        }
        if (min2 == 4001) return min1;
        else if (array[min1 - min] != array[min2 - min]) return min1;
        return min2;
    }
}

File: 백준/Silver/2108. 통계학/통계학.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

#define FAST_IO cin.tie(NULL)->ios::sync_with_stdio(false);

using namespace std;

#define REP(i,a,b) for (int i = a; i <= b; i++)

int max_, min_, sum_;
vector<int> numbers;

int mode() {
    int k = max_ - min_ + 1;
    vector<int> arr(k);
    if (numbers.size() == 1) return numbers[0];
    for (int num : numbers) {
        arr[num - min_]++;
    }

    int min1 = 4001, min2 = 4001, frequency = 1;
    REP(i, 0, k - 1) {
        if (arr[i] > frequency) {
            frequency = arr[i];
            min1 = i + min_;
            min2 = 4001;
        } else if (arr[i] == frequency) {
            if (i + min_ < min1) {
                min2 = min1;
                min1 = i + min_;
            }
            else if (i + min_ < min2) min2 = i + min_;
        }
    }
    if (min2 == 4001) return min1;
    else if (arr[min1 - min_] != arr[min2 - min_]) return min1;
    return min2;
}

int main() {
    FAST_IO

    int n; cin >> n;
    max_ = -4000;
    min_ = 4000;
    sum_ = 0;
    numbers.resize(n);
    REP(i, 0, n - 1) {
        cin >> numbers[i];
        sum_ += numbers[i];
        max_ = max(max_, numbers[i]);
        min_ = min(min_, numbers[i]);
    }

    cout << lround(sum_ / (double) n) << '\n';
    sort(numbers.begin(), numbers.end());
    cout << numbers[n / 2] << '\n';
    cout << mode() << '\n';
    cout << max_ - min_ << '\n';

    return EXIT_SUCCESS;
}